src/connectWithShell.tsx

```diff
@@ -118,28 +118,79 @@ export interface ConnectWithShellOptions {
+            'To fix this, call connectWithShell() from Entry Point attach() or extend(). ' +
+            'If you really have to create this component type dynamically, ' +
+            'either pass {allowOutOfEntryPoint:true} in options, or use shell.runLateInitializer().'
+        //TODO: replace with throw after a grace period
+        boundShell.log.warning(errorText)
+    }
+}
+
 export function connectWithShell<S = {}, OP = {}, SP = {}, DP = {}>(
     mapStateToProps: MapStateToProps<S, OP, SP>,
     mapDispatchToProps: MapDispatchToProps<OP, DP>,
     boundShell: Shell,
     options: ConnectWithShellOptions = {}
 ) {
-    const validateLifecycle = (component: React.ComponentType<any>) => {
-        if (boundShell.wasInitializationCompleted() && !options.allowOutOfEntryPoint) {
-            const componentText = component.displayName || component.name || component
-            const errorText =
-                `connectWithShell(${boundShell.name})(${componentText}): ` +
-                'attempt to create component type outside of Entry Point lifecycle. ' +
-                'To fix this, call connectWithShell() from Entry Point attach() or extend(). ' +
-                'If you really have to create this component type dynamically, ' +
-                'either pass {allowOutOfEntryPoint:true} in options, or use shell.runLateInitializer().'
-            //TODO: replace with throw after a grace period
-            boundShell.log.warning(errorText)
+    return (component: React.ComponentType<OP & SP & DP>) => {
+        validateComponentLifecycle(boundShell, options, component)
+        return wrapWithShellContext(component, mapStateToProps, mapDispatchToProps, boundShell, options)
+    }
+}
+
+export function withObservers<OP>(
+    boundShell: Shell,
+    observers: ChangeObserver[],
+    connectedComponent: React.ComponentType<OP>
+) {
+    type ObservableWrapperState = { observedVersion: number };
+    class ObservableWrapperComponent extends React.Component<OP, ObservableWrapperState> {
+        public connectedComponent: React.ComponentType<OP>
+
+        constructor(props: OP) {
+            super(props)
+            this.connectedComponent = connectedComponent;
+            this.setState({ observedVersion: 1 })
+
+            observers.forEach(observer => observer.subscribe(boundShell, () => {
+                this.setState({ observedVersion: this.state.observedVersion + 1 })
+            }));
         }
+
+        public render() {
+            const Component = this.connectedComponent
+            return <Component {...this.props} observedVersion={this.state.observedVersion} />
+        }
+    }
+
+
+    const hoc: React.FunctionComponent<OP> = (props) => {
+        return (
+            <ObservableWrapperComponent {...props} />
+        )
     }
+    return hoc;
+} 
 
+export function connectWithShellAndObserve<S = {}, OP = {}, SP = {}, DP = {}>(
+    observers: ChangeObserver[],
```

Comment:
I see a big gap here - 
A consumer of API that wants to connect, needs to know about implementation of a specific state of a provider, and state can be implicitly observable or broadcasted.

I think a state being observable and not broadcasted needs to be reflected in the contract between modules, so that is state changes notification type, it will not be backward compatible.

What I suggest is that when a state is observable - the observer will be the only way to access the state, so wrapping with with-observer HOC will be also the actual provider of the props, that than implicit request to an API.

WDYT?

Reply:
Yep, that's an issue I also thought about. If an API function returns data from observable state, and it's called from some components' mapStateToProps, the developer of the component has no clue that he/she should include some observer in the connect. Which is bad.

The problem was that observers cannot pass their observed state to subscribers - this would be a leak of package internal state; the state will become part of the public API interface.

Having said that, it's good that you stretch the issue. We have to go an extra mile and do this:

- make subscriber receive the updated data in a parameter (which brings us back to the conventional RX interface). But - it won't be the internal state. Instead, it will be some transformation of the state into a public interface. 
- API functions that return data from observable state, must be moved to an observable interface - an object that can be only accessed through an observer.

```
interface Observable<T> {
  getValue(): T
  subscribe(listener: (nextValue:T)=>void): Unsubscribe
}

// before
interface MyAPI {
  getSomething(): Something
  getSomethingElse(): SomethingElse
}

// after
interface SomethingElseSelectors {
  getSomethingElse(): SomethingElse
}
interface MyAPI {
  getSomething(): Something
  observables: {
    somethingElse: Observable<SomethingElseSelectors>
  }
}
```

LMK if it looks good to you now